msgpack/field.go

```diff
@@ -74,6 +74,8 @@ func collectFields(fields []*field, t reflect.Type, visited map[reflect.Type]boo
 		}
 
 		if len(index) == d {
+			_ = fields[len(fields)-1] // bounds check hint to compiler
```

Comment:
Can you elaborate on that? I've seen bounds checks (namely the std lib) to guard a specific length of a slice, but not ever like this

Reply:
This commit is very old so I don't remember that, but IIRC scope is `for i := 0; i < len(fields); i++`.

Reply:
yeah, but when do you expect `len(fields)` to go out of bounds? Usually I've seen bound checks like these guard a slice of unknown length when we're explicitly accessing certain indexes like this:

```go
func someFunc(s []string) {
  _ = s[1]
  s[0] = "hey"
  s[1] = "test!"
}
```

While what you've done here is you're not guarding an explicit length like: `_ = fields[3]` but guarding the size which is restricted by the language itself, that is: the last index has to be `len(slice)-1`, there's no other way. It seems completely redundant to me and confusing.

Am I missing something? :)

Reply:
I'll re-detect where is BCE optimized using `-gcflags='-d=ssa/check_bce/debug=3'`.

But in context of language specifications, the `fields` value is not array, actual type is slice. It means, the gc compiler (still can't optimization? I don't know) might be imagine will growing slice. And actually we also can grow `fields` slice after the BCE'ed place.
e.g., `encoding/binary` package is `return` byte slice immediately after the BCE place, but this msgpack packages logic is used for loop after the BCE place, therefore works BCE optimization. (above is my current imagine. I'll verify again that the BCE is working properly when I have time anyway.)

Reply:
interesting!

Still, `fields` is not local to the `collect` function and might be manipulated after you do the bounds check so you won't really eliminate the BCE done on `fields[j] = fields[i]`, I've verified that with this whole file:

- before manual bounds check:

```shell
$ go run -gcflags="-d=ssa/check_bce/debug=1" field.go
# command-line-arguments
./field.go:83:12: Found IsInBounds
./field.go:87:19: Found IsSliceInBounds
./field.go:100:10: Found IsInBounds
```

- after:

```shell
└> go run -gcflags="-d=ssa/check_bce/debug=1" field.go
# command-line-arguments
./field.go:78:17: Found IsInBounds
./field.go:84:12: Found IsInBounds
./field.go:88:19: Found IsSliceInBounds
./field.go:101:10: Found IsInBounds
```

You can see you've just introduced an extra check there and haven't got rid of the ones performed in the loop.

Correct me If I'm wrong and thanks a lot for answering, this is quiet interesting! :)